### ISSUE TITLE
scala.pickling.shareNothing._ required to avoid java.lang.ClassCastException

### ISSUE THREAD

**Issue by simbo1905** [scala/pickling#310]
Using version 0.10.0: 

```
lazy val root = (project in file(".")).
  settings(
        libraryDependencies ++= Seq(
          "org.scala-lang.modules" %% "scala-pickling" % "0.10.0"
          )
  )
```

Using it with `sbt run` to compile this file: 

```
case class ProposalNumber(counter: Int, nodeIdentifier: Int) {
  def >(that: ProposalNumber) = if (this == that) false else if (this.counter > that.counter) true else if (this.counter < that.counter) false else this.nodeIdentifier > that.nodeIdentifier
  def >=(that: ProposalNumber) = if (this == that) true else if (this.counter > that.counter) true else if (this.counter < that.counter) false else this.nodeIdentifier > that.nodeIdentifier
  def <(that: ProposalNumber) = if (this == that) false else if (this.counter > that.counter) false else if (this.counter < that.counter) true else this.nodeIdentifier < that.nodeIdentifier
  def <=(that: ProposalNumber) = if (this == that) true else if (this.counter > that.counter) false else if (this.counter < that.counter) true else this.nodeIdentifier < that.nodeIdentifier
  override def toString = f"N(c=${counter.toLong-Int.MinValue},n=$nodeIdentifier)"
}

case class Identifier(val from: Int, val number: ProposalNumber, val logIndex: Long) {
  override def toString = f"I(f=$from,n=$number,s=$logIndex)"
}

case class Progress(val highestPromised: ProposalNumber, val highestCommitted: Identifier){
  override def toString = s"P(p=$highestPromised,c=$highestCommitted)"
}

object Hi { 
    def main(args: Array[String]) {
import scala.pickling.Defaults._, scala.pickling.json._
      val minValue = ProposalNumber(Int.MinValue, Int.MinValue)
      val minBookwork = Progress(minValue, Identifier(1, minValue, 0))
      val bytes = minBookwork.pickle.value
      val parsed = bytes.unpickle[Progress]
      if( parsed != minBookwork ) {
        println("broken :(")
      } else {
        println("FTW! :)")
      }
    } 
}
```

Gets a runtime ClassCastException: 

```
[error] (run-main-0) java.lang.ClassCastException: Identifier cannot be cast to ProposalNumber
java.lang.ClassCastException: Identifier cannot be cast to ProposalNumber
    at Hi$ProgressUnpickler1$2$IdentifierUnpickler1$2$.unpickle(hw.scala:24)
    at Hi$ProgressUnpickler1$2$.unpickle(hw.scala:24)
    at scala.pickling.Unpickler$class.unpickleEntry(Pickler.scala:79)
    at Hi$ProgressUnpickler1$2$.unpickleEntry(hw.scala:24)
    at scala.pickling.functions$.unpickle(functions.scala:11)
    at scala.pickling.UnpickleOps.unpickle(Ops.scala:23)
    at Hi$.main(hw.scala:24)
    at Hi.main(hw.scala)
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke(Method.java:483)
```

This can be fixed by adding `scala.pickling.shareNothing._` to the imports to be: 

```
import scala.pickling.Defaults._,scala.pickling.json._
 import scala.pickling.shareNothing._ // sharing json bug on 0.10.0
```

**Comment by eed3si9n** [comment]
I was able to reproduce this from REPL using 0.10.0:

``` scala
scala> case class ProposalNumber(counter: Int, nodeIdentifier: Int) {
     |   def >(that: ProposalNumber) = if (this == that) false else if (this.counter > that.counter) true else if (this.counter < that.counter) false else this.nodeIdentifier > that.nodeIdentifier
     |   def >=(that: ProposalNumber) = if (this == that) true else if (this.counter > that.counter) true else if (this.counter < that.counter) false else this.nodeIdentifier > that.nodeIdentifier
     |   def <(that: ProposalNumber) = if (this == that) false else if (this.counter > that.counter) false else if (this.counter < that.counter) true else this.nodeIdentifier < that.nodeIdentifier
     |   def <=(that: ProposalNumber) = if (this == that) true else if (this.counter > that.counter) false else if (this.counter < that.counter) true else this.nodeIdentifier < that.nodeIdentifier
     |   override def toString = f"N(c=${counter.toLong-Int.MinValue},n=$nodeIdentifier)"
     | }
defined class ProposalNumber

scala> case class Identifier(val from: Int, val number: ProposalNumber, val logIndex: Long) {
     |   override def toString = f"I(f=$from,n=$number,s=$logIndex)"
     | }
defined class Identifier

scala> case class Progress(val highestPromised: ProposalNumber, val highestCommitted: Identifier){
     |   override def toString = s"P(p=$highestPromised,c=$highestCommitted)"
     | }
defined class Progress

scala> import scala.pickling.Defaults._, scala.pickling.json._
import scala.pickling.Defaults._
import scala.pickling.json._

scala> val minValue = ProposalNumber(Int.MinValue, Int.MinValue)
minValue: ProposalNumber = N(c=0,n=-2147483648)

scala> val minBookwork = Progress(minValue, Identifier(1, minValue, 0))
minBookwork: Progress = P(p=N(c=0,n=-2147483648),c=I(f=1,n=N(c=0,n=-2147483648),s=0))

scala> val bytes = minBookwork.pickle.value
bytes: String =
{
  "$type": "Progress",
  "highestPromised": {
    "counter": -2147483648,
    "nodeIdentifier": -2147483648
  },
  "highestCommitted": {
    "from": 1,
    "logIndex": "0",
    "number": { "$ref": 1 }
  }
}

scala> val parsed = bytes.unpickle[Progress]
java.lang.ClassCastException: Identifier cannot be cast to ProposalNumber
  at $line5$readiwProgressUnpickler$macro$10$2$$line4$readiwIdentifierUnpickler$macro$20$2$.unpickle(<console>:22)
  at $line5$readiwProgressUnpickler$macro$10$2$.unpickle(<console>:22)
  at scala.pickling.Unpickler$class.unpickleEntry(Pickler.scala:79)
  at $line5$readiwProgressUnpickler$macro$10$2$.unpickleEntry(<console>:22)
  at scala.pickling.functions$.unpickle(functions.scala:11)
  at scala.pickling.UnpickleOps.unpickle(Ops.scala:23)
  ... 43 elided
```

**Comment by jsuereth** [comment]
This is now fixed in 0.11.x with the new macros.

**Comment by phaller** [comment]
@jsuereth Should we add a regression test (on 0.11.x) for this?
